lib/python/dtype.cpp

```diff
@@ -95,7 +90,10 @@ objects containing binned data. They cannot be used directly to create arrays of
            dtype<core::bin<Dataset>>,
            dtype<python::PyObject>,
        }) {
-    std::cout << i++ << '\n';
+    // For some reason OSX causes trouble when we return `t` directly, with the
+    // dtype not matching the one in the registry (possibly related to problems
+    // of symbols across library boundaries). Instead, we find by name and
+    // return the key.
```

Comment:
The old implementation used a lambda that captured `t` by ref and stored it in pybind11's storage. So the ref outlived the initializer list in the for loop above. Maybe on OSX, that list contains local copies but on Ubuntu and Windows it contains refs to the static `dtype` objects?
Did you try capturing `t` by value?

Reply:
Hmm, good point, I will try.